graphbolt/src/shared_memory_utils.cc

```diff
@@ -18,164 +18,193 @@
-          read_from_archive(archive, "tensor_" + std::to_string(i) + "_shape")
-              .toIntVector();
-      auto dtype =
-          read_from_archive(archive, "tensor_" + std::to_string(i) + "_dtype")
-              .toScalarType();
-      metas.push_back({true, shape, dtype});
-    } else {
-      metas.push_back({false, {}, torch::ScalarType::Undefined});
+    auto key =
+        read_from_archive(archive, std::string("key_") + std::to_string(i))
+            .toStringRef();
+    auto tensor = this->ReadTorchTensor();
+    tensor_dict.insert(key, tensor.value());
+  }
+  return tensor_dict;
+}
+
+void SharedMemoryHelper::WriteTorchArchiveInternal(
+    torch::serialize::OutputArchive& archive) {
+  std::stringstream serialized;
+  archive.save_to(serialized);
+  auto serialized_str = serialized.str();
+  auto meta_ptr = this->GetCurrentMetaPtr();
+  static_cast<int64_t*>(meta_ptr)[0] = serialized_str.size();
+  memcpy(
+      static_cast<char*>(meta_ptr) + sizeof(int64_t), serialized_str.data(),
+      serialized_str.size());
+  int64_t rounded_size = GetRoundedSize(serialized_str.size());
+  this->MoveMetaPtr(sizeof(int64_t) + rounded_size);
+}
+
+void SharedMemoryHelper::WriteTorchTensorInternal(
+    torch::optional<torch::Tensor> tensor) {
+  if (tensor.has_value()) {
+    size_t memory_size = tensor.value().numel() * tensor.value().element_size();
+    auto data_ptr = this->GetCurrentDataPtr();
+    auto contiguous_tensor = tensor.value().contiguous();
+    memcpy(data_ptr, contiguous_tensor.data_ptr(), memory_size);
+    this->MoveDataPtr(GetRoundedSize(memory_size));
+  }
+}
+
+void SharedMemoryHelper::Flush() {
+  // The first archive records the size of the tensor data.
+  torch::serialize::OutputArchive first_archive;
+  size_t data_size = 0;
+  for (auto tensor : tensors_to_write_) {
+    if (tensor.has_value()) {
+      auto tensor_size = tensor.value().numel() * tensor.value().element_size();
+      data_size += GetRoundedSize(tensor_size);
     }
   }
-  return std::tuple_cat(
-      std::forward_as_tuple(std::move(meta_shm)),
-      LoadTensorsDataFromSharedMemory(name + "_data", metas));
+  first_archive.write("data_size", static_cast<int64_t>(data_size));
+  meta_shared_memory_ =
+      std::make_unique<SharedMemory>(GetSharedMemoryMetaName(name_));
+  meta_shared_memory_->Create(max_meta_memory_size_);
+  meta_offset_ = 0;
+  this->WriteTorchArchiveInternal(first_archive);
+  for (auto& archive : meta_to_write_) {
+    this->WriteTorchArchiveInternal(archive);
+  }
+  data_shared_memory_ =
+      std::make_unique<SharedMemory>(GetSharedMemoryDataName(name_));
+  data_shared_memory_->Create(data_size);
+  data_offset_ = 0;
+  for (auto tensor : tensors_to_write_) {
+    this->WriteTorchTensorInternal(tensor);
+  }
+  meta_to_write_.clear();
+  tensors_to_write_.clear();
+
+  meta_offset_ = 0;
+  // Skip the first archive recording data size before read.
+  this->ReadTorchArchive();
+  data_offset_ = 0;
```

Comment:
There are 4 cleanup and 1 read in this section, does the order matter?

Reply:
L194-195 remove the storages of data that need to write to shared memory. L197 and L200 reset the offsets for later read operations. L199 removes the first archive, which cannot be seen by the user of this class. So the only dependency is that we need to reset the metadata offset before reading the first torch archive.